### ISSUE TITLE
Constrained `TypeVar`s can be resolved to subclasses of constraints

### ISSUE THREAD

**Issue by decorator-factory** [microsoft/pyright#3010]
According to PEP 484, if a type variable is constrained, it can only be assigned to the explicitly listed types, not to any of the subtypes.

However, it seems like this restriction doesn't work when explicitly specifying the type in brackets.

```py
from typing import TypeVar, Generic

class A: ...
class B: ...
class A2(A): ...
T = TypeVar("T", A, B)

class F(Generic[T]):
    def __init__(self, thing: T) -> None:
        self.thing = thing

f1 = F(A2())
reveal_type(f1)  # F[A], which is correct
reveal_type(f1.thing)  # A, which is correct

f2 = F[A2](A2())
reveal_type(f2)  # F[A2], incorrect
reveal_type(f2.thing)  # A2, incorrect
```
Another example would be with a contravariant type variable:
```py
from typing import TypeVar, Generic

class A: ...
class B: ...
class A2(A): ...

T = TypeVar("T", A, B, contravariant=True)

class F(Generic[T]):
    def __init__(self, thing: Callable[[T], None]) -> None:
        self.thing = thing

def callback(a: A) -> None:
    ...

f1: F[A2] = F(callback)
reveal_type(f1)  # F[A2]
reveal_type(f1.thing)  # (A2) -> None
```
Here, the type of `F(callback)` is inferred as `F[A]` correctly, but the explicit annotation makes `f1` have the type `F[A2]`. 

Shouldn't `F[A2]` be prohibited in general? Or am I misunderstanding PEP 484?

**Comment by decorator-factory** [comment]
After thinking about it a little bit, I am not entirely sure if a contravariant constrained type variable makes sense :thinking:

**Comment by erictraut** [comment]
You're misunderstanding PEP 484. When a constrained type variable is used, it's permissible for a subtype to be bound to it. However, the "solution" to the type is the original type rather than the subtype.

```python
T = TypeVar("T", float, str)

def func(x: T) -> T: ...

reveal_type(func(3.0)) # float
reveal_type(func(3)) # float, not int
```

And...

```python
from typing import TypeVar

class A: ...
class B: ...
class A2(A): ...

T1 = TypeVar("T1", A, B)

def func1(x: T1) -> T1: ...

T2 = TypeVar("T2")

def func2(x: T2) -> T2: ...

reveal_type(func1(A2())) # A
reveal_type(func2(A2())) # A2
```

**Comment by decorator-factory** [comment]
I understand that. But in my code example, the "solution" to the type is actually the subtype if you specify it explicitly:
```py
f2 = F[A2](A2())
reveal_type(f2)  # F[A2]
reveal_type(f2.thing)  # A2
```

**Comment by erictraut** [comment]
Ah, I see what you mean. Yeah, that might be an issue. I'll look into it more.

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.221, which I just published. It will also be included in the next release of pylance.
